pkg/catalogsourceconfig/handler.go

```diff
@@ -0,0 +1,143 @@
+package catalogsourceconfig
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	opsrc "github.com/operator-framework/operator-marketplace/pkg/operatorsource"
+
+	olm "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	"github.com/operator-framework/operator-marketplace/pkg/apis/marketplace/v1alpha1"
+	"github.com/operator-framework/operator-sdk/pkg/sdk"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Copied from https://github.com/operator-framework/operator-lifecycle-manager/blob/master/pkg/controller/registry/configmap_loader.go#L18
+// TBD: Vendor in the folder once we require more than just these constants from the OLM registry code
+const (
+	ConfigMapCRDName     = "customResourceDefinitions"
+	ConfigMapCSVName     = "clusterServiceVersions"
+	ConfigMapPackageName = "packages"
+)
+
+// Handler is the interface that wraps the Handle method
+type Handler interface {
+	Handle(ctx context.Context, event sdk.Event) error
+}
+
+type handler struct {
+	reader opsrc.DatastoreReader
+}
+
+// Handle handles a new event associated with CatalogSourceConfig type
+func (h *handler) Handle(ctx context.Context, event sdk.Event) error {
+	csc := event.Object.(*v1alpha1.CatalogSourceConfig)
+	// Ignore the delete event as the garbage collector will clean up the created resources as per the OwnerReference
+	if event.Deleted {
+		logrus.Infof("Deleted %s CatalogSourceConfig in %s namespace", csc.Name, csc.Spec.TargetNamespace)
+		return nil
+	}
+	data, err := h.createCatalogData(csc)
+	if err != nil {
+		return err
+	}
+	return createCatalogSource(csc, data)
+}
+
+// createCatalogData constructs the ConfigMap data by reading the manifest information of all packages
+// from the datasource
+func (h *handler) createCatalogData(csc *v1alpha1.CatalogSourceConfig) (map[string]string, error) {
+	packageIds := getPackageIds(csc.Spec.Packages)
+	data := make(map[string]string)
+	if len(packageIds) < 1 {
+		return data, fmt.Errorf("No packages specified in CatalogSourceConfig %s/%s", csc.Namespace, csc.Name)
+	}
+
+	// TBD: Do we create a CatalogSource per package?
+	for id := range packageIds {
+		manifest, err := h.reader.Read(packageIds[id])
+		if err != nil {
+			logrus.Errorf("Error \"%v\" getting manifest for package ID %s", err, packageIds[id])
+			continue
+		}
+		// TODO: Add more error checking
```

Comment:
Not sure why you are trying to pull these variables from OLM?

Reply:
Because the ConfigMap's data section has to be constructed in the format mandated by OLM.